### PR TITLE
NoUnneededNestedControlBlockFixer - introduction

### PR DESCRIPTION
**Pull request by SpacePossum**
```
$ php php-cs-fixer describe no_unneeded_nested_control_block
Description of no_unneeded_nested_control_block rule.
There must not be unneeded nested `if-if` and `else-if` blocks.
```

```diff
--- Original
+++ New
@@ -1,5 +1,5 @@
 <?php
-if ($a) {
-    if ($b) {
-    }
+if ($a && $b) {
+
+
 }
```

```diff
--- Original
+++ New
@@ -1,7 +1,7 @@
 <?php
 if ($a) {

-} else {
-    if ($b) {
-    }
+} elseif ($b) {
+
+
 }
```

### CODE REVIEW

**Comment by julienfalque** [comment]
IMO this can be considered an enhancement rather than a new feature (no opt-in), WDYT?

**Comment by SpacePossum** [comment]
the fixer claims it fixes following PSR and this case is not in the PSR's... so I can see both sides,
as it point to master I can also make it opt-out, maybe that is better

**Comment by julienfalque** [comment]
Ah I didn't notice you were modifying the `ElseifFixer`. In this case an opt-in option makes sense to me indeed.

But this change fits better in `NoUselessElseFixer`: in the case an `else` block only contains an `if` block, the code can be simplified to an `elseif` block (or `else if` which would then be replaced with a `elseif` by `EleseifFixer`), e.g.:

```diff
 if ($foo) {
     // foo
-} else {
-    if ($bar) {
-        // bar
-    }
+} else if ($bar) {
+    // bar
 }
```
Supporting this scenario in `NoUselessElseFixer` can be done without opt-in option IMO.

**Comment by SpacePossum** [comment]
The `NoUselessElseFixer` removes `else` where not needed. In this scenario  however we change;
 `if (x) {} else { if (x) {} }` 
into;
 `if (x) {} elseif (x) {} ` 

Since we are not removing an `else` statement but transforming it to `elseif` I think is the responsibility of the `ElseifFixer`.

**Comment by julienfalque** [comment]
You're right, this should not be done by `NoUselessElseFixer`, but I still feel it should not be done by `ElseifFixer` either. `ElseifFixer` basically replaces `else if` with `elseif`, here we want to replace `else`+nested `if` with `else if`, which are similar but different responsibilities.

**Comment by SpacePossum** [comment]
Im thinking a new fixer `no_unneeded_if_nesting` that also fixes
```
if (a) {
    if (b) {
        if (c) {
             d();
        }
    }
}
```

into
```
if (a && b && c) { 
     d(); 
}
```

WDYT?

**Comment by julienfalque** [comment]
Even better :)

File: tests/Fixer/ControlStructure/NoUnneededNestedControlBlockFixerTest.php [hunk 1]
```diff
@@ -0,0 +1,577 @@
...
+    }
+}
+',
+        ];
+
+        yield 'multiple "else" case' => [
+            '<?php
+if ($foo) {
+
+} else {
```
Comment (julienfalque) [inline tests/Fixer/ControlStructure/NoUnneededNestedControlBlockFixerTest.php:354]: The remaining `else` blocks are unneeded and should be removed as well.

File: tests/Fixer/ControlStructure/NoUnneededNestedControlBlockFixerTest.php [hunk 2]
```diff
@@ -0,0 +1,577 @@
...
+    }
+}
+',
+        ];
+
+        yield 'multiple "elseif" case' => [
+            '<?php
+if ($foo) {
+
+} elseif($c && $bar) {
```
Comment (julienfalque) [inline tests/Fixer/ControlStructure/NoUnneededNestedControlBlockFixerTest.php:389]: The remaining `elseif` blocks are unneeded and should be removed as well.

File: tests/Fixer/ControlStructure/NoUnneededNestedControlBlockFixerTest.php [hunk 3]
```diff
@@ -0,0 +1,577 @@
...
+            '<?php
+
+if ($foo) {
+    if ($bar) {
+
+    }
+} elseif ($a) {
+    echo 1;
+}',
+        ];
```
Comment (julienfalque) [inline tests/Fixer/ControlStructure/NoUnneededNestedControlBlockFixerTest.php:502]: Please add the following test case (maybe using other assignation operators as well):
```php
        yield 'variable assignations in parent blocks' => [
            '<?php
if ((null !== $foo = foo()) && (null !== $bar = bar()) && null !== $baz = baz()) {
            echo 1;
}',
            '<?php
if (null !== $foo = foo()) {
    if (null !== $bar = bar()) {
        if (null !== $baz = baz()) {
            echo 1;
        }
    }
}',
        ];
```

File: tests/Fixer/ControlStructure/NoUnneededNestedControlBlockFixerTest.php (hunk 2 above)
Comment (SpacePossum) [reply tests/Fixer/ControlStructure/NoUnneededNestedControlBlockFixerTest.php:389]: That would change the code behavior and make the fixer risky, also it out of scope of the rule.
For example;
```php
<?php

function foo() {
    echo 1;
}

if (false) {

} elseif (foo()) {

}
```
you reason all code can be removed here, yet `foo()` won't be called than and now it is

Comment (SpacePossum) [reply tests/Fixer/ControlStructure/NoUnneededNestedControlBlockFixerTest.php:389]: Besides, having empty conditional block is a pretty rare use case and is typically covered by SCA already. Could be a nice new fixer though to remove such blocks.

Comment (julienfalque) [reply tests/Fixer/ControlStructure/NoUnneededNestedControlBlockFixerTest.php:389]: Right, my point is invalid regarding `elseif` but I think it's still valid for `else`.

File: tests/Fixer/ControlStructure/NoUnneededNestedControlBlockFixerTest.php (hunk 1 above)
Comment (SpacePossum) [reply tests/Fixer/ControlStructure/NoUnneededNestedControlBlockFixerTest.php:354]: The remaining ones are not nested, so out of scope. We already have a fixer for the useless else and a priority check for it. So I think it is not needed to remove the else block here.

Comment (julienfalque) [reply tests/Fixer/ControlStructure/NoUnneededNestedControlBlockFixerTest.php:354]: > The remaining ones are not nested

I don't understand. The expected part of this test case does have `if`s inside `else`s, and [the test case at line 105](https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/4799/files#diff-357a383ebe0f37632dbc43479ea5661cR105) implies they should be replaced with `elseif`s.

The following:
```php
if ($a) {
} elseif ($b) {
} elseif ($c) {
}
```
is the same as
```php
if ($a) {
} else {
    if ($b) {
    } else {
        if ($c) {
        }
    }
}
```
right?

Also I don't think any other fixer would take care of that (`no_useless_else` would only remove `else` if the previous `if` has e.g. a `return` statement).

File: tests/Fixtures/Integration/priority/no_unneeded_nested_control_block,no_useless_else.test [hunk 4]
```diff
@@ -0,0 +1,23 @@
+--TEST--
+Integration of fixers: no_unneeded_nested_control_block,no_useless_else.
+--RULESET--
+{"no_unneeded_nested_control_block": true, "no_useless_else": true}
+--EXPECT--
+<?php
+if (1) {
+    return 1;
+} elseif (2) {
```
[the remark this answers is not in the document]
Comment (SpacePossum) [reply tests/Fixtures/Integration/priority/no_unneeded_nested_control_block,no_useless_else.test:9]: The current priority (values) and fixers (like `braces`) makes this not possible.
We do get good end results though.
When using `no_unneeded_nested_control_block`, `no_useless_else` and also `no_superfluous_elseif` it still works as expected.

File: tests/Fixer/ControlStructure/NoUnneededNestedControlBlockFixerTest.php (hunk 1 above)
Comment (SpacePossum) [reply tests/Fixer/ControlStructure/NoUnneededNestedControlBlockFixerTest.php:354]: > `no_useless_else` would only remove else if the previous if has e.g. a return statement

It shouldn't, so if you have an example that fails please share it :)
For example:
```
   1) test.php (no_useless_else)
      ---------- begin diff ----------
--- Original
+++ New
@@ @@
 <?php
 if (1) {
     echo 123;
-} else {
+}  
 
-}
+
```

I'll look into the other cases to see what can be improved :+1:

**Comment by keradus** [comment]
I'm dropping the milestone, as PR is marked as a draft

**Pull request closed without merging**